### DISCUSSION TITLE
Consistent pluralization of routes?

### DISCUSSION THREAD

**Discussion by redbar0n** [vikejs/vike#1337] [Discussion]
Re: https://github.com/vikejs/vike/issues/1322#issuecomment-1842614526  and https://vike.dev/file-structure :

> ```shell
> # Route: /products
> /pages/products+Page.js
> # Route: /product/@id
> /pages/product/@id+Page.js
> ```
> 
> Looks weird but it's functional.

Maybe it's a good idea to have consistent/uniform pluralization throughout the framework?

Though I'm not entirely sure how hard-coded pluralization of routes is in the framework (or if it's entirely user-defined), or if it's just the documentation that is a bit inconsistent. See:

https://github.com/search?q=repo%3Avikejs%2Fvike+%22products%22&type=code
vs
https://github.com/search?q=repo%3Avikejs%2Fvike+%22%2Fproduct%22&type=code

In any case, I suggest the following.

Either pluralize all url's:

```shell
/products
/products/@id
```

Or have them all in singular:

```shell
/product
/product/@id
```

The less variability the better, since:
- Uniform resource URI's makes them more predictable.
- Pluralization was one of the more [unintuitive / constant-stumble-block](https://www.reddit.com/r/rails/comments/8uk1k0/whats_your_opinion_on_rails_inflection_naming/) features of Rails, imho. It's super at first, because it reads so well. But as you scale (and run into the edge cases and have to deal with the variability) it becomes annoying.
- `products` and `product` are strictly speaking not two distinct resources, but they are two distinct URI's. But `an application should expose many URIs, one for each resource;` https://en.wikipedia.org/wiki/Resource-oriented_architecture

Arguments for pluralized:
- `/products/@id` can be seen as accessing the list of products and getting one with a specific id.
- `/products` is more logical than `/product` when you want to access the list of products. Unless you interpret the last one as giving you a product index `/product/index.html`.
- Plural is more common in REST style API's?

Arguments for singular:
- Aligns better with:
  - DB tables, as you typically write things like `SELECT name FROM user`.
  - Class names.
- Some names are awkward to pluralize, like 
- Combined resources (like join tables) opens the slight uncertainty if all parts of the name should be pluralized or not (e.g. `UsersProducts`).
- [Some plural nouns are irregular](https://stackoverflow.com/a/42562934/380607) (unpredictable) in English. Possible endings: `-s`, `-es`, `-i`, `-a`, or just inconvenient to pluralize.

It would be harder to change this convention later. V1 design is a good time, imho.

Some related discussions:

https://stackoverflow.com/questions/6845772/should-i-use-singular-or-plural-name-convention-for-rest-resources

https://stackoverflow.com/questions/18570158/handling-singular-and-plural-controllers-routes

https://www.reddit.com/r/AskProgramming/comments/15rxwfd/singular_or_plural_nouns/

https://github.com/kubernetes/kubernetes/issues/18622

**Comment by brillout** [comment]
> hard-coded pluralization of routes is in the framework

It isn't hardcoded into Vike; the user has full freedom to choose any URL structure he wants.

I like the user friendliness of `/products` for a list of products and `/product/@id` for a specific product, that's why the docs showcase these. But the user can do whatever he wants.
